Fix effective sample size estimation

junpenglao opened the pull request:
Close #2810
Reference implementation in PyStan: https://github.com/stan-dev/pystan/pull/415
* improve autocorr and autocov function, use fft convolve for autocorrelation computation

Commit:
Fix effective sample size estimation

Close #2810
Reference implementation in PyStan: https://github.com/stan-dev/pystan/pull/415
- improve autocorr and autocov function, use fft convolve for autocorrelation computation

ColCarroll commented on pymc3/stats.py:
```diff
@@ -59,44 +60,60 @@ def wrapped_f(pymc3_obj, *args, **kwargs):
+    n = len(y)
+    result = fftconvolve(y, y[::-1])
+    acorr = result[len(result) // 2:]
+    acorr /= (n - np.arange(n))
```
super minor: `np.arange(n, 0, -1)` is about 2x as fast (~1µs vs 2µs for `n=1000`), but also modestly more readable.  (I guess `fftconvolve` takes the longest of this stuff?)

ColCarroll commented on pymc3/stats.py:
```diff
@@ -59,44 +60,60 @@ def wrapped_f(pymc3_obj, *args, **kwargs):
     """
-    S = autocov(x, lag)
-    return S[0, 1] / np.sqrt(np.prod(np.diag(S)))
+    y = x - x.mean()
```
is this why we don't `x -= x.mean()`? maxim caught me doing that in `imcmc` last week...

junpenglao replied:
Yep, took me awhile to find out because it silently modifies the trace... the computed value is always off by a little bit but some is super close - drives me crazy for a day :-/

Commit:
Improve effective sample size implementation

Effective sample size computed using autocov and Geyer's initial monotone sequence criterion for vhat.

Commit:
Improve effective sample size implementation

Effective sample size computed using autocov and Geyer's initial monotone sequence criterion for vhat.

Co-authored-by: Junpeng Lao <junpeng.lao@unifr.ch>
Co-authored-by: Sharan Yalburgi <sharanyalburgi@gmail.com>

junpenglao commented:
Numerical validation could be found in https://gist.github.com/junpenglao/aa6ea70d55d1c93fc11e45e3bf8666bd#file-current_implementation_check-ipynb

There are small differences with Stan due to the implementation of autocov.

also cc @sharanry who started the work ;-)

ColCarroll commented:
This might be apt here to share credit:

https://github.com/blog/2496-commit-together-with-co-authors

Commit:
Merge remote-tracking branch 'origin/improve_eff_n' into improve_eff_n

Commit:
Turn off automatic convergent check for short chain

Commit:
update release note

Commit:
Turn off convergent check in benchmark model

To have more accurate computation of the things we want to benchmark. 
effective_n computation is slower than before for large number of samples (at 10000 sample it is around 10x slower).
There should be a large increase in effective_n benchmark.

junpenglao commented:
@ColCarroll I edited the benchmark suite, since there might be some side effect of this PR:
- Setting `progressbar=False, compute_convergence_checks=False` To have a more accurate computation of the things we want to benchmark. 
- effective_n computation is slower than before for large number of samples (at 10000 samples it is around 10x slower).
- There should be a large improvement in the effective_n benchmark.

ColCarroll commented:
:shipit:

ColCarroll commented:
Benchmarks don't run on travis, so I'm merging now, and will watch those over the next day or two.
